### PR TITLE
Add the way to get search stat information when 'set shortmess-=S'

### PR DESCRIPTION
**Pull request by tyru**
As of [8.1.1270](https://github.com/vim/vim/commit/9dfa3139198b38b28673e251a3756430065914e9), vim can show current match position and count like `[1/20]` at command-line.
It is useful but cannot get the status from Vim script, without using `execute(':unsilent :norm! n')` as test script does.
But I don't want to execute `n` to know the status, because it throws CursorMoved, and changes jumplist, and so on.

~So I add a `v:searchstat` variable which can be referenced by Vim script without running any command.~

---

**EDIT:** added a function (`searchcount()`) instead of a variable.

### CODE REVIEW

**Comment by tyru** [comment]
# Usecase

* Show the status in statusline.

# Problem

> As of [8.1.1270](https://github.com/vim/vim/commit/9dfa3139198b38b28673e251a3756430065914e9), vim can show current match position and count like `[1/20]` at command-line.
> It is useful but cannot get the status from Vim script, without using `execute(':unsilent :norm! n')` as test script does.
> But I don't want to execute `n` to know the status, because it throws CursorMoved, and changes jumplist, and so on.

Current specification has another problem.
If user has mappings of `n`, `N`,

```vim
nnoremap n nzz
nnoremap N Nzz
```

`zz` redraws the screen and `[1/20]` message disappears.

If user can get the status by variable, user can :echo the status at any timing.

```vim
set shortmess-=S

" zz removes search status information
"nnoremap n nzz
"nnoremap N Nzz

nnoremap n nzz:echo printf('[%d/%d]', v:searchstat.current, v:searchstat.count)<CR>
nnoremap N Nzz:echo printf('[%d/%d]', v:searchstat.current, v:searchstat.count)<CR>
```

**Comment by mattn** [comment]
Please add test for empty status. Also I'm interesting whether the stat must be reset somewhere.

**Comment by chrisbra** [comment]
I think this needs some `ifdef FEAT_EVAL` for the tiny builds. Also I'd also be interested how long the v:searchstat variable is valid and when it will be reset.

**Comment by tyru** [comment]
Okay. >empty status, ifdef

Currently no change (cleared, updated) is made until the next `n`, `N`, `/`, ... commands run AND `shortmess` does not contain `S`.

I'm also wondering that

* when the variable is changed (updated, cleared)
* how long the variable is valid

Personally I think

* `set shortmess+=S` should make the variable empty
* otherwise, the variable does not change until the next `n`, `N`, `/`, ... commands

How do you think?

**Comment by brammool** [comment]
Takuya Fujiwara wrote:

> # Usecase
> 
> * Show the status in statusline.

I wonder, if you show the stats in the statusline, shouldn't there be a
way to not show them on the command line?

v:searchstat should probably be cleared when editing another file.
Perhaps also when switching to another window.

> # Problem
> 
> > As of [8.1.1270](https://github.com/vim/vim/commit/9dfa3139198b38b28673e251a3756430065914e9), vim can show current match position and count like `[1/20]` at command-line.
> > It is useful but cannot get the status from Vim script, without using `execute(':unsilent :norm! n')` as test script does.
> > But I don't want to execute `n` to know the status, because it throws CursorMoved, and changes jumplist, and so on.
> 
> Current specification has another problem.
> If user has mappings of `n`, `N`,
> 
> ```vim
> nnoremap n nzz
> nnoremap N Nzz
> ```
> 
> `zz` redraws the screen and `[1/20]` message disappears.
> 
> If user can get the status by variable, user can :echo the status at
> any timing.
> 
> ```vim
> set shortmess-=S
> 
> " zz removes search status information
> "nnoremap n nzz
> "nnoremap N Nzz
> 
> nnoremap n nzz:echo printf('[%d/%d]', v:searchstat.current, v:searchstat.count)<CR>
> nnoremap N Nzz:echo printf('[%d/%d]', v:searchstat.current, v:searchstat.count)<CR>
> ```

Add to the help somewhere?

-- 
Never overestimate a man's ability to underestimate a woman.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
> I wonder, if you show the stats in the statusline, shouldn't there be a way to not show them on the command line?

After thinking some more about the use case, I also wondered about that. Once we have `v:searchstat`, I guess many people wouldn't want to have it displayed in the command line anymore, but let the plugins make use of it. Probably would need another option.

**Comment by tyru** [comment]
> I wonder, if you show the stats in the statusline, shouldn't there be a way to not show them on the command line?

Ah, that's true.

> v:searchstat should probably be cleared when editing another file. Perhaps also when switching to another window.

Yes. That's right.

> Probably would need another option.

You mean another interface (function, option, ...) is better?
I think a variable is good to make use of the stat because it's simple, and fast to reference the stat from statusline (function is a little bit slow?).
But please tell me if you have any idea.

**Comment by brammool** [comment]
> > I wonder, if you show the stats in the statusline, shouldn't there be a way to not show them on the command line?
> 
> Ah, that's true.
> 
> > v:searchstat should probably be cleared when editing another file. Perhaps also when switching to another window.
> 
> Yes. That's right.
> 
> > Probably would need another option.
> 
> You mean another interface (function, option, ...) is better?
> I think a variable is good to make use of the stat because it's simple, and fast to reference the stat from statusline (function is a little bit slow?).
> But please tell me if you have any idea.

What could work:
- Make v:searchstat available independend from the "S" flag in
  'shortmess'
- When the value of v:searchstat is requested, compute the values, like
  when displaying it in the command line.
- Keep track of when the values in v:searchstat become invalid (when the
  text in the command line would be updated, when editing another
  buffer, when the search string changes, etc.)

That way the "S" flag in 'shortmess' just controls whether the stats are
displayed in the command line.

The value is cached to avoid recomputing it when nothing changed, and we
don't compute it when it is not used.

-- 
"Oh, no!  NOT the Spanish Inquisition!"
"NOBODY expects the Spanish Inquisition!!!"
				-- Monty Python sketch --
"Oh, no!  NOT another option!"
"EVERYBODY expects another option!!!"
				-- Discussion in vim-dev mailing list --

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
How about to add new identify of the status for statusline? ex `%t`
And add new function to get values of the identiy. ex `getstatuslineinfo("%t")` ?

**Comment by tyru** [comment]
> How about to add new identify of the status for statusline? ex `%t`
> And add new function to get values of the identiy. ex `getstatuslineinfo("%t")` ?

It also has lifecycle issue yet.

If statusline has `%t`, the `[1/20]` stat appears in current window.
There should be many options when it disappears (after n seconds, next search command run, ...).

And in most cases, users want to see the stat only in statusline of current window.
But Vim searches all statuslines of current tab page.

---

So, I want to simply add a function to get the stat.
Sorry for confusing @brammool and @chrisbra,
but just adding a function is simple and avoids lifecycle issue,
users / plugins can use it and determine when to delete the display, and so on.

One problem is the performance.
searching word count may become slow even max count is 99.

So the function should be able to get the cached value when `shortmess` does not contain `S`.
In the latest commit, `searchcount()` was added and if the arguments were omitted, it can get the cached value.
(Please forgive horrible code now...)

I wrote also PoC API in help:

```diff
diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
index 15ffed343..f8b7baafa 100644
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -2564,6 +2564,7 @@ screenrow()			Number	current cursor row
 screenstring({row}, {col})	String	characters at screen position
 search({pattern} [, {flags} [, {stopline} [, {timeout}]]])
 				Number	search for {pattern}
+searchcount([{dict}])		Dictionary	the search stats
 searchdecl({name} [, {global} [, {thisblock}]])
 				Number	search for variable declaration
 searchpair({start}, {middle}, {end} [, {flags} [, {skip} [...]]])
@@ -7910,6 +7911,31 @@ search({pattern} [, {flags} [, {stopline} [, {timeout}]]])	*search()*
 		The 'n' flag tells the function not to move the cursor.
 
 
+searchcount({dict})					*searchcount()*
+		The result is almost the same structure as {dict}.
+		Some keys can be specified by only arguments.
+		This function changes {dict} in-place and returns it.
+		If the keys of {dict} are omitted, default values were used.
+
+		 key		type	arg only	meaning ~
+		  dir		String	no	(default: "/" if |v:searchforward| is
+					no	non-zero, otherwise "?")
+		  lastpos	List	no	`[lnum, col, off]` . see |getpos()|
+					no	(default: the last position if
+					no	|'shortmess'| does not contain "S",
+					no	otherwise `[0, 0, 0]`)
+		  pattern	String	no	(default: |@/|)
+		  bufnr		Number	no	(default: current buffer)
+		  changedtick	Number	no	(default: |b:changedtick|)
+		  current	Number	no
+		  count		Number	no
+		  wrap		Number	no
+		  maxcount	Number	yes	Max number of count (default: 99)
+		  recompute	Number	yes	if specified and zero, the
+						last computed values when
+						|'shortmess'| has not "S" flag
+						(default: non-zero)
+
 searchdecl({name} [, {global} [, {thisblock}]])			*searchdecl()*
 		Search for the declaration of {name}.
 
```

**Comment by tyru** [comment]
To get the cached value (when `set shortmess-=S`):

```vim
:echo searchcount({'recompute': 0})
{'bufnr': 1, 'pattern': '\<Vim\>', 'dir': '/', 'wrap': 0, 'changedtick': 4, 'lastpos': [65, 17, 0], 'current': 3, 'recompute': 0, 'count': 10}
```

**Comment by brammool** [comment]
A function avoids the need for tricks with updating the variable, I like that.
The function you define is rather complex. Why not just return a dict with the info, without passing in anything?  And I don't see a good reason to also include things already available in another way.

**Comment by brammool** [comment]
Are you planning to make the suggested changes?

**Comment by tyru** [comment]
sorry, I missed that.
I'll do it.

**Comment by tyru** [comment]
> Why not just return a dict with the info, without passing in anything? And I don't see a good reason to also include things already available in another way.

Done it.
Sorry for not responding for a long time.
Please see the help and tests for the detailed function behavior.

@brammool 
I also want to add options to change the max count & timeout of `n`, `N` (`SEARCH_STAT_DEF_MAX_COUNT`, `SEARCH_STAT_DEF_TIMEOUT`).
How do you think?

**Comment by tyru** [comment]
I also found the way to show search stat information on command-line
even `n` and `N` were mapped to `nzz`, `Nzz` and `zz` clears command-line (I did this).

```vim
nnoremap n nzz
nnoremap N Nzz
```

This is little hacky thing so that I hesitated to add this to vim help.
I publish the diff here however.

```diff
diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index f65d46fec..5e698c80e 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -6738,6 +6738,26 @@ A jump table for the options with a short description can be found at |Q_op|.
 	NOTE: This option is set to the Vi default value when 'compatible' is
 	set and to the Vim default value when 'compatible' is reset.
 
+							*shortmess-S-faq*
+	When you set `set shortmess-=S`, if you want to show the count message
+	even when you have mapped to |n| and |N| like this: >
+
+		" Centering screen after n, N.
+		" But zz also clears search count message!
+		nnoremap n nzz
+		nnoremap N Nzz
+<
+	You can do like this: >
+
+		nnoremap <silent> <SID>(cmd:n) :<C-u>let w:search_stat = execute('unsilent norm! n')->split('\n')->get(-1, '')<CR>
+		nnoremap <silent> <SID>(cmd:N) :<C-u>let w:search_stat = execute('unsilent norm! N')->split('\n')->get(-1, '')<CR>
+		nnoremap <silent> <SID>(echo-search-stat) :<C-u>redraw <Bar> echo '' <Bar> echo w:search_stat<CR>
+
+		nmap <script> n <SID>(cmd:n)zz<SID>(echo-search-stat)
+		nmap <script> N <SID>(cmd:N)zz<SID>(echo-search-stat)
+<
+	Or you can use |searchcount()| function for more further information.
+
 				 *'shortname'* *'sn'* *'noshortname'* *'nosn'*
 'shortname' 'sn'	boolean	(default off)
 			local to buffer
```

**Comment by tyru** [comment]
I also wrote detailed example to show the search stat in statusline. c4100e115d1e9eece700bb7f24fb4f3385282224

**Comment by tyru** [comment]
Sorry for bothering of `git push -f` many times.

* Added more example: 729a21517f72cf671c75c02c4d67cd87048795a6
* Change default value of `recompute`: bbacdd9f4dc9ace97c1ea46cc0610ca20bb086e3

**Comment by tyru** [comment]
* 61f6881ba3a1f0a52155d0229cf2be4f99864d66: Support `exact_match` of return value (could not know if cursor is on the matched text, or just after cursor position)
* 25da18fa525cdf5f68206d2117c73b604b9d3974: Rename `timeout` to `incomplete`. and support to check if search exceeded max count
  * `incomplete == 0`: search was fully completed
  * `incomplete == 1`: recomputing was timed out
  * `incomplete == 2`: max count exceeded

**Comment by tyru** [comment]
93e5e7913bd04bd0dc10f8d2fe9f7d0c34e051c4: Fixed the build was broken. And add `max_count` property of return value (see example for help)

@brammool This should be the final for my second proposal. Please take a look if you are interested.

**Comment by brammool** [comment]
This function looks useful, but I'm not sure about the details.  Let me include it as-is now, and we can further improve it after that.

**Comment by tyru** [comment]
Thanks!

**Comment by lacygoill** [comment]
I don't know whether it can be optimized, but the longer the line the slower `searchcount()` seems to be:
```vim
vim9script
@/ = 'x'
repeat('x', 100000)->setline(1)
nno <expr> n Func()
def g:Func(): string
    searchcount({'maxcount': 1000, 'timeout': 500})
    return 'n'
enddef
```
Keep pressing `n` for a few seconds, then stop: Vim still needs several seconds to process the keypresses.

This is not merely caused by the size of the buffer or the number of matches.  Here is the exact same text, but splitted on 1000 lines:
```vim
vim9script
@/ = 'x'
range(1000)->map({i -> repeat('x', 100)->setline(i + 1)})
nno <expr> n Func()
def g:Func(): string
    searchcount({'maxcount': 1000, 'timeout': 500})
    return 'n'
enddef
```
Keep pressing `n` for a few seconds, then stop: Vim stops immediately.

The issue is really the length of the line.

**Comment by brammool** [comment]
This will require profiling to find out what the time is spent on.
